Network/DNS/Transport.hs

```diff
@@ -232,3 +256,15 @@ isIllegal dom
   | any (\x -> BS.length x > 63)
         (BS.split '.' dom)      = True
   | otherwise                   = False
+
+repeatUntilM :: Monad m => (a -> Bool) -> m a -> m [a]
+repeatUntilM predicate x = go []
+  where go acc = do
+          val <- x
+          let acc' = val:acc
+          if predicate val
+            then return $ reverse acc'
+            else go acc'
+
+isAXFRLast :: DNSMessage -> Bool
+isAXFRLast = (== SOA) . rrtype . last . answer
```

Comment:
AXFR should present a monadic data stream, not a pure list.  If the caller then chooses to buffer the data in memory, that's the caller's choice, but the API should present a constant-space interface.